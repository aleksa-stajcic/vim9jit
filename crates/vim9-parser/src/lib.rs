#![allow(dead_code)]

use anyhow::Result;
use vim9_lexer::new_lexer;
use vim9_lexer::Lexer;
use vim9_lexer::Token;
use vim9_lexer::TokenKind;

#[derive(Debug)]
pub struct Program {
    commands: Vec<ExCommand>,
}

#[derive(Debug, PartialEq)]
pub enum ExCommand {
    Vim9Script(Vim9ScriptCommand),
    Var(VarCommand),
    Return(ReturnCommand),
    Def(DefCommand),
    If(IfCommand),
    Call(CallCommand),
    Finish(Token),

    Skip,
    EndOfFile,
    Comment(Token),
    NoOp(Token),
}

#[derive(Debug, PartialEq)]
pub struct Vim9ScriptCommand {
    noclear: bool,
    eol: Token,
}

#[derive(Debug, PartialEq)]
pub struct VarCommand {
    var: Token,
    pub name: Name,
    equal: Token,
    pub expr: Expression,
    eol: Token,
}

impl VarCommand {
    pub fn parse(parser: &mut Parser) -> Result<ExCommand> {
        let var = VarCommand {
            var: parser.expect_token_with_text(TokenKind::Identifier, "var")?,
            name: Name::parse(parser)?,
            // TODO: Type Hints
            equal: parser.expect_token(TokenKind::Equal)?,
            expr: Expression::parse(parser)?,
            eol: parser.expect_eol()?,
        };

        Ok(ExCommand::Var(var))
    }
}

#[derive(Debug, PartialEq)]
pub struct Body {
    pub commands: Vec<ExCommand>,
}

impl Body {
    pub fn parse_until(parser: &mut Parser, identifier: &str) -> Result<Body> {
        let mut commands = vec![];
        while parser.current_token.text != identifier {
            commands.push(parser.parse_command()?);
        }

        Ok(Body { commands })
    }
}

#[derive(Debug, PartialEq)]
pub struct IfCommand {
    if_tok: Token,
    pub condition: Expression,
    if_eol: Token,
    pub body: Body,
    endif_tok: Token,
    endif_eol: Token,
}

impl IfCommand {
    pub fn parser(parser: &mut Parser) -> Result<ExCommand> {
        Ok(ExCommand::If(IfCommand {
            if_tok: parser.expect_identifier_with_text("if")?,
            condition: Expression::parse(parser)?,
            if_eol: parser.expect_eol()?,
            body: Body::parse_until(parser, "endif")?,
            endif_tok: parser.expect_identifier_with_text("endif")?,
            endif_eol: parser.expect_eol()?,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub struct DefCommand {
    def: Token,
    pub name: Name,
    pub args: Signature,
    pub ret: Option<Type>,
    def_eol: Token,
    pub body: Body,
    enddef: Token,
    end_eol: Token,
}

impl DefCommand {
    pub fn parse(parser: &mut Parser) -> Result<ExCommand> {
        Ok(ExCommand::Def(DefCommand {
            def: parser.expect_identifier_with_text("def")?,
            name: Name::parse(parser)?,
            args: Signature::parse(parser)?,
            ret: {
                if parser.current_token.kind == TokenKind::Colon {
                    Some(Type {
                        colon: parser.expect_token(TokenKind::Colon)?,
                        name: Name::parse(parser)?,
                    })
                } else {
                    None
                }
            },
            def_eol: parser.expect_eol()?,
            body: Body::parse_until(parser, "enddef")?,
            enddef: parser.expect_identifier_with_text("enddef")?,
            end_eol: parser.expect_eol()?,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub struct CallCommand {
    call: Option<Token>,
    pub name: Name,
    pub args: Arguments,
    eol: Token,
}

impl CallCommand {
    pub fn parse(parser: &mut Parser) -> Result<ExCommand> {
        Ok(ExCommand::Call(CallCommand {
            call: None,
            name: Name::parse(parser)?,
            args: Arguments::parse(parser)?,
            eol: parser.expect_eol()?,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub struct Arguments {
    open: Token,
    pub args: Vec<Expression>,
    close: Token,
}

impl Arguments {
    pub fn parse(parser: &mut Parser) -> Result<Arguments> {
        Ok(Arguments {
            open: parser.expect_token(TokenKind::LeftParen)?,
            args: {
                let mut args = vec![];
                while parser.current_token.kind != TokenKind::RightParen {
                    args.push(Expression::parse(parser)?);

                    // TODO: Consume ','?
                }

                args
            },
            close: parser.expect_token(TokenKind::RightParen)?,
        })
    }
}

#[derive(Debug, PartialEq)]
pub struct Type {
    colon: Token,
    pub name: Name,
}

#[derive(Debug, PartialEq)]
pub struct Signature {
    open: Token,
    close: Token,
}

impl Signature {
    fn parse(parser: &mut Parser) -> Result<Signature> {
        Ok(Self {
            open: parser.expect_token(TokenKind::LeftParen)?,
            close: parser.expect_token(TokenKind::RightParen)?,
        })
    }
}

#[derive(Debug, PartialEq)]
pub struct Parameter {
    pub name: Name,
    colon: Token,
    // pub typ: ,
}

#[derive(Debug, PartialEq)]
pub struct ReturnCommand {
    ret: Token,
    pub expr: Expression,
    eol: Token,
}

impl ReturnCommand {
    pub fn parse(parser: &mut Parser) -> Result<ExCommand> {
        Ok(ExCommand::Return(Self {
            ret: parser.expect_token_with_text(TokenKind::Identifier, "return")?,
            expr: Expression::parse(parser)?,
            eol: parser.expect_eol()?,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub enum Name {
    Identifier(RawIdentifier),
    ScopedIdentifier,
}

impl Name {
    fn parse(parser: &mut Parser) -> Result<Name> {
        // Todo: Other names
        Ok(Name::Identifier(RawIdentifier {
            name: parser.expect_token(TokenKind::Identifier)?.text,
        }))
    }
}

#[derive(Debug, PartialEq)]
pub enum Expression {
    Name,
    Number(Number),
    String(VimString),
}

impl Expression {
    pub fn parse(parser: &mut Parser) -> Result<Expression> {
        let expr = match &parser.current_token.kind {
            TokenKind::Integer => Expression::Number(Number {
                value: parser.current_token.text.clone(),
            }),
            TokenKind::DoubleQuoteString => {
                Expression::String(VimString::DoubleQuote(parser.current_token.text.clone()))
            }
            TokenKind::SingleQuoteString => {
                Expression::String(VimString::SingleQuote(parser.current_token.text.clone()))
            }

            _ => todo!("expresson parse: {:?}", parser.current_token),
        };

        parser.next_token();
        Ok(expr)
    }
}

#[derive(Debug, PartialEq)]
pub enum VimString {
    SingleQuote(String),
    DoubleQuote(String),
}

#[derive(Debug, PartialEq)]
pub struct RawIdentifier {
    name: String,
}

#[derive(Debug, PartialEq)]
pub struct Number {
    // ??? should this be a number?
    value: String,
}

#[derive(Debug)]
pub struct Parser {
    lexer: Lexer,
    current_token: Token,
    peek_token: Token,
}

impl Parser {
    pub fn new(mut lexer: Lexer) -> Self {
        Self {
            current_token: lexer.next_token(),
            peek_token: lexer.next_token(),
            lexer,
        }
    }

    pub fn expect_eol(&mut self) -> Result<Token> {
        self.expect_token(TokenKind::EndOfLine)
    }

    pub fn expect_token(&mut self, kind: TokenKind) -> Result<Token> {
        let token = self.current_token.clone();
        if token.kind != kind {
            return Err(anyhow::anyhow!("Got token: {:?}, Expected: {:?}", token, kind));
        }

        self.next_token();
        Ok(token)
    }

    pub fn expect_token_with_text(&mut self, kind: TokenKind, text: &str) -> Result<Token> {
        let token = self.current_token.clone();
        if token.kind != kind {
            panic!("Got token: {:?}, Expected kind: {:?}", token, kind);
            return Err(anyhow::anyhow!("Got token: {:?}, Expected kind: {:?}", token, kind));
        }

        if token.text != text {
            panic!("Got token: {:?}, Expected text: {:?}", token, text);
            return Err(anyhow::anyhow!("Got token: {:?}, Expected text: {:?}", token, text));
        }

        self.next_token();
        Ok(token)
    }

    pub fn peek_identifier_with_text(&self, text: &str) -> bool {
        self.peek_token.kind == TokenKind::Identifier && self.peek_token.text == text
    }

    pub fn expect_identifier_with_text(&mut self, text: &str) -> Result<Token> {
        self.expect_token_with_text(TokenKind::Identifier, text)
    }

    pub fn next_token(&mut self) -> Token {
        self.current_token = self.peek_token.clone();
        self.peek_token = self.lexer.next_token();

        self.current_token.clone()
    }

    fn command_match(&self, full: &str) -> bool {
        self.current_token.text == full
    }

    pub fn parse_command(&mut self) -> Result<ExCommand> {
        // For the following branches, you need to return early if it completely consumes
        // the last character and advances past.
        //
        // This is the desired behavior for `parse` which will consume until the end of line
        // generally speaking.
        let ex = match &self.current_token.kind {
            TokenKind::EndOfFile => panic!("EndOfFile!"),
            TokenKind::Comment => ExCommand::Comment(self.current_token.clone()),
            TokenKind::EndOfLine => ExCommand::NoOp(self.current_token.clone()),
            TokenKind::Identifier => {
                if self.command_match("vim9script") {
                    self.next_token();
                    ExCommand::Vim9Script(Vim9ScriptCommand {
                        noclear: if self.current_token.kind == TokenKind::EndOfLine {
                            false
                        } else {
                            self.expect_identifier_with_text("noclear")?;
                            true
                        },
                        eol: self.expect_eol()?,
                    })
                } else if self.command_match("var") {
                    return Ok(VarCommand::parse(self)?);
                } else if self.command_match("def") {
                    return Ok(DefCommand::parse(self)?);
                } else if self.command_match("return") {
                    return Ok(ReturnCommand::parse(self)?);
                } else if self.command_match("if") {
                    return Ok(IfCommand::parser(self)?);
                } else if self.command_match("finish") {
                    ExCommand::Finish(self.current_token.clone())
                } else {
                    if self.peek_token.kind == TokenKind::LeftParen {
                        return Ok(CallCommand::parse(self)?);
                    } else {
                        ExCommand::NoOp(self.current_token.clone())
                    }
                }
            }
            _ => ExCommand::NoOp(self.current_token.clone()),
        };

        self.next_token();
        Ok(ex)
    }

    pub fn parse_program(&mut self) -> Program {
        let mut program = Program { commands: vec![] };

        while self.current_token.kind != TokenKind::EndOfFile {
            let command = self.parse_command().unwrap();
            if command != ExCommand::Skip {
                program.commands.push(command);
            }
        }

        program
    }
}

fn snapshot_parsing(input: &str) -> String {
    let lexer = new_lexer(input);
    let mut parser = Parser::new(lexer);
    let program = parser.parse_program();

    format!("{:#?}", program.commands)
}

#[cfg(test)]
mod test {
    use super::*;

    macro_rules! snapshot {
        ($name:tt, $path:tt) => {
            #[test]
            fn $name() {
                let contents = include_str!($path);
                let mut settings = insta::Settings::clone_current();
                settings.set_snapshot_path("../testdata/output/");
                settings.bind(|| {
                    insta::assert_snapshot!(snapshot_parsing(contents));
                });
            }
        };
    }

    snapshot!(test_var, "../testdata/snapshots/simple_var.vim");
    snapshot!(test_ret, "../testdata/snapshots/simple_ret.vim");
    snapshot!(test_comment, "../testdata/snapshots/comment.vim");
    snapshot!(test_header, "../testdata/snapshots/header.vim");

    // TODO: Slowly but surely, we can work towards this
    // snapshot!(test_matchparen, "../testdata/snapshots/matchparen.vim");
}
