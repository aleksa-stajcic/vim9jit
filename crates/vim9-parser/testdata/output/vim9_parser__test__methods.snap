---
source: crates/vim9-parser/src/lib.rs
assertion_line: 1962
expression: snapshot_parsing(contents)
---
[
    Vim9Script(
        Vim9ScriptCommand {
            noclear: false,
            eol: Token(EndOfLine, "\n", (0,10)->(0,10)),
        },
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (2,0)->(2,3)),
            ty: None,
            name: Raw(s),
            equal: Token(Equal, "=", (2,6)->(2,7)),
            expr: String(
                DoubleQuote(
                    "hello",
                ),
            ),
            eol: Token(EndOfLine, "\n", (2,15)->(2,15)),
        },
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (3,0)->(3,3)),
            ty: None,
            name: Raw(l),
            equal: Token(Equal, "=", (3,6)->(3,7)),
            expr: MethodCall(
                MethodCall {
                    left: Identifier(
                        Raw(s),
                    ),
                    tok: Token(MethodArrow, "->", (3,9)->(3,11)),
                    right: f: Identifier(Raw(len)) arg: [],
                },
            ),
            eol: Token(EndOfLine, "\n", (3,16)->(3,16)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (4,0)->(4,0)),
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (5,0)->(5,3)),
            ty: None,
            name: Raw(l2),
            equal: Token(Equal, "=", (5,7)->(5,8)),
            expr: MethodCall(
                MethodCall {
                    left: String(
                        DoubleQuote(
                            "hello",
                        ),
                    ),
                    tok: Token(MethodArrow, "->", (5,16)->(5,18)),
                    right: f: Identifier(Raw(len)) arg: [],
                },
            ),
            eol: Token(EndOfLine, "\n", (5,23)->(5,23)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (6,0)->(6,0)),
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (7,0)->(7,3)),
            ty: None,
            name: Raw(x),
            equal: Token(Equal, "=", (7,6)->(7,7)),
            expr: MethodCall(
                MethodCall {
                    left: MethodCall(
                        MethodCall {
                            left: MethodCall(
                                MethodCall {
                                    left: MethodCall(
                                        MethodCall {
                                            left: Array(
                                                ArrayLiteral {
                                                    open: Token(LeftBracket, "[", (7,8)->(7,9)),
                                                    elements: [
                                                        Number(
                                                            VimNumber {
                                                                value: "5",
                                                            },
                                                        ),
                                                        Number(
                                                            VimNumber {
                                                                value: "4",
                                                            },
                                                        ),
                                                        Number(
                                                            VimNumber {
                                                                value: "2",
                                                            },
                                                        ),
                                                        Number(
                                                            VimNumber {
                                                                value: "1",
                                                            },
                                                        ),
                                                        Number(
                                                            VimNumber {
                                                                value: "7",
                                                            },
                                                        ),
                                                        Number(
                                                            VimNumber {
                                                                value: "12",
                                                            },
                                                        ),
                                                        Number(
                                                            VimNumber {
                                                                value: "8",
                                                            },
                                                        ),
                                                    ],
                                                    close: Token(RightBracket, "]", (7,29)->(7,30)),
                                                },
                                            ),
                                            tok: Token(MethodArrow, "->", (7,30)->(7,32)),
                                            right: f: Identifier(Raw(sort)) arg: [],
                                        },
                                    ),
                                    tok: Token(MethodArrow, "->", (8,5)->(8,7)),
                                    right: f: Identifier(Raw(filter)) arg: [
                                        Lambda(
                                            Lambda {
                                                args: Signature {
                                                    open: Token(LeftParen, "(", (8,14)->(8,15)),
                                                    params: [
                                                        Parameter {
                                                            name: Raw(_),
                                                            ty: None,
                                                            equal: None,
                                                            default_val: None,
                                                        },
                                                        Parameter {
                                                            name: Raw(x),
                                                            ty: None,
                                                            equal: None,
                                                            default_val: None,
                                                        },
                                                    ],
                                                    close: Token(RightParen, ")", (8,19)->(8,20)),
                                                },
                                                ret: None,
                                                arrow: Token(Arrow, "=>", (8,21)->(8,23)),
                                                body: Body {
                                                    commands: [
                                                        Return(
                                                            ReturnCommand {
                                                                ret: Token(Virtual, "", (0,0)->(0,0)),
                                                                expr: Some(
                                                                    Infix(
                                                                        InfixExpression {
                                                                            token: Token(EqualTo, "==", (8,30)->(8,32)),
                                                                            operator: EqualTo,
                                                                            left: Infix(
                                                                                InfixExpression {
                                                                                    token: Token(Percent, "%", (8,26)->(8,27)),
                                                                                    operator: Modulo,
                                                                                    left: Identifier(
                                                                                        Raw(x),
                                                                                    ),
                                                                                    right: Number(
                                                                                        VimNumber {
                                                                                            value: "2",
                                                                                        },
                                                                                    ),
                                                                                },
                                                                            ),
                                                                            right: Number(
                                                                                VimNumber {
                                                                                    value: "0",
                                                                                },
                                                                            ),
                                                                        },
                                                                    ),
                                                                ),
                                                                eol: Token(Virtual, "", (0,0)->(0,0)),
                                                            },
                                                        ),
                                                    ],
                                                },
                                            },
                                        ),
                                    ],
                                },
                            ),
                            tok: Token(MethodArrow, "->", (9,5)->(9,7)),
                            right: f: Identifier(Raw(map)) arg: [
                                Lambda(
                                    Lambda {
                                        args: Signature {
                                            open: Token(LeftParen, "(", (9,11)->(9,12)),
                                            params: [
                                                Parameter {
                                                    name: Raw(_),
                                                    ty: None,
                                                    equal: None,
                                                    default_val: None,
                                                },
                                                Parameter {
                                                    name: Raw(y),
                                                    ty: None,
                                                    equal: None,
                                                    default_val: None,
                                                },
                                            ],
                                            close: Token(RightParen, ")", (9,16)->(9,17)),
                                        },
                                        ret: None,
                                        arrow: Token(Arrow, "=>", (9,18)->(9,20)),
                                        body: Body {
                                            commands: [
                                                Return(
                                                    ReturnCommand {
                                                        ret: Token(Virtual, "", (0,0)->(0,0)),
                                                        expr: Some(
                                                            Infix(
                                                                InfixExpression {
                                                                    token: Token(Plus, "+", (9,23)->(9,24)),
                                                                    operator: Plus,
                                                                    left: Identifier(
                                                                        Raw(y),
                                                                    ),
                                                                    right: Number(
                                                                        VimNumber {
                                                                            value: "1",
                                                                        },
                                                                    ),
                                                                },
                                                            ),
                                                        ),
                                                        eol: Token(Virtual, "", (0,0)->(0,0)),
                                                    },
                                                ),
                                            ],
                                        },
                                    },
                                ),
                            ],
                        },
                    ),
                    tok: Token(MethodArrow, "->", (10,5)->(10,7)),
                    right: f: Identifier(Raw(sort)) arg: [],
                },
            ),
            eol: Token(EndOfLine, "\n", (10,13)->(10,13)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (11,0)->(11,0)),
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (12,0)->(12,3)),
            ty: None,
            name: Raw(expr_prec),
            equal: Token(Equal, "=", (12,14)->(12,15)),
            expr: MethodCall(
                MethodCall {
                    left: Prefix(
                        PrefixExpression {
                            token: Token(Minus, "-", (12,16)->(12,17)),
                            operator: Minus,
                            right: Number(
                                VimNumber {
                                    value: "1.234",
                                },
                            ),
                        },
                    ),
                    tok: Token(MethodArrow, "->", (12,22)->(12,24)),
                    right: f: Identifier(Raw(string)) arg: [],
                },
            ),
            eol: Token(EndOfLine, "\n", (12,32)->(12,32)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (13,0)->(13,0)),
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (14,0)->(14,3)),
            ty: None,
            name: Raw(foo),
            equal: Token(Equal, "=", (14,8)->(14,9)),
            expr: MethodCall(
                MethodCall {
                    left: Identifier(
                        Raw(base),
                    ),
                    tok: Token(MethodArrow, "->", (14,14)->(14,16)),
                    right: f: Identifier(Raw(name)) arg: [
                        Identifier(
                            Raw(args),
                        ),
                    ],
                },
            ),
            eol: Token(EndOfLine, "\n", (14,26)->(14,26)),
        },
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (15,0)->(15,3)),
            ty: None,
            name: Raw(foo),
            equal: Token(Equal, "=", (15,8)->(15,9)),
            expr: MethodCall(
                MethodCall {
                    left: Identifier(
                        Raw(base),
                    ),
                    tok: Token(MethodArrow, "->", (15,14)->(15,16)),
                    right: f: DictAccess(DictAccess { container: Identifier(Raw(some)), dot: Token(Dot, ".", (15,20)->(15,21)), index: RawIdentifier { name: "name" } }) arg: [
                        Identifier(
                            Raw(args),
                        ),
                    ],
                },
            ),
            eol: Token(EndOfLine, "\n", (15,31)->(15,31)),
        },
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (16,0)->(16,3)),
            ty: None,
            name: Raw(foo),
            equal: Token(Equal, "=", (16,8)->(16,9)),
            expr: MethodCall(
                MethodCall {
                    left: Identifier(
                        Raw(base),
                    ),
                    tok: Token(MethodArrow, "->", (16,14)->(16,16)),
                    right: f: Index(IndexExpression { container: Identifier(Raw(alist)), open: Token(LeftBracket, "[", (16,21)->(16,22)), index: Item(Identifier(Raw(idx))), close: Token(RightBracket, "]", (16,25)->(16,26)) }) arg: [
                        Identifier(
                            Raw(args),
                        ),
                    ],
                },
            ),
            eol: Token(EndOfLine, "\n", (16,32)->(16,32)),
        },
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (17,0)->(17,3)),
            ty: None,
            name: Raw(foo),
            equal: Token(Equal, "=", (17,8)->(17,9)),
            expr: MethodCall(
                MethodCall {
                    left: Identifier(
                        Raw(base),
                    ),
                    tok: Token(MethodArrow, "->", (17,14)->(17,16)),
                    right: f: Grouped(GroupedExpression { open: Token(LeftParen, "(", (17,16)->(17,17)), expr: Call(f: Identifier(Raw(getFuncRef)) arg: []), close: Token(RightParen, ")", (17,29)->(17,30)) }) arg: [
                        Identifier(
                            Raw(args),
                        ),
                    ],
                },
            ),
            eol: Token(EndOfLine, "\n", (17,36)->(17,36)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (18,0)->(18,0)),
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (19,0)->(19,3)),
            ty: None,
            name: Raw(x),
            equal: Token(Equal, "=", (19,6)->(19,7)),
            expr: Array(
                ArrayLiteral {
                    open: Token(LeftBracket, "[", (19,8)->(19,9)),
                    elements: [
                        Number(
                            VimNumber {
                                value: "1",
                            },
                        ),
                        Number(
                            VimNumber {
                                value: "4",
                            },
                        ),
                        Number(
                            VimNumber {
                                value: "2",
                            },
                        ),
                        Number(
                            VimNumber {
                                value: "5",
                            },
                        ),
                    ],
                    close: Token(RightBracket, "]", (19,19)->(19,20)),
                },
            ),
            eol: Token(EndOfLine, "\n", (19,20)->(19,20)),
        },
    ),
    Eval(
        EvalCommand {
            eval: None,
            expr: MethodCall(
                MethodCall {
                    left: Identifier(
                        Raw(x),
                    ),
                    tok: Token(MethodArrow, "->", (20,1)->(20,3)),
                    right: f: Identifier(Raw(sort)) arg: [],
                },
            ),
            eol: Token(EndOfLine, "\n", (20,9)->(20,9)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (21,0)->(21,0)),
    ),
    Var(
        VarCommand {
            var: Token(Identifier, "var", (22,0)->(22,3)),
            ty: None,
            name: Raw(foo),
            equal: Token(Equal, "=", (22,8)->(22,9)),
            expr: MethodCall(
                MethodCall {
                    left: Array(
                        ArrayLiteral {
                            open: Token(LeftBracket, "[", (22,10)->(22,11)),
                            elements: [
                                Identifier(
                                    Raw(slnum),
                                ),
                                Identifier(
                                    Raw(scol),
                                ),
                            ],
                            close: Token(RightBracket, "]", (22,22)->(22,23)),
                        },
                    ),
                    tok: Token(MethodArrow, "->", (22,23)->(22,25)),
                    right: f: Identifier(Raw(cursor)) arg: [],
                },
            ),
            eol: Token(EndOfLine, "\n", (22,33)->(22,33)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (23,0)->(23,0)),
    ),
    Comment(
        Token(Comment, "# Testing", (24,0)->(24,9)),
    ),
    Eval(
        EvalCommand {
            eval: None,
            expr: MethodCall(
                MethodCall {
                    left: Array(
                        ArrayLiteral {
                            open: Token(LeftBracket, "[", (25,0)->(25,1)),
                            elements: [
                                Identifier(
                                    Raw(slnum),
                                ),
                                Identifier(
                                    Raw(scol),
                                ),
                            ],
                            close: Token(RightBracket, "]", (25,12)->(25,13)),
                        },
                    ),
                    tok: Token(MethodArrow, "->", (25,13)->(25,15)),
                    right: f: Identifier(Raw(cursor)) arg: [],
                },
            ),
            eol: Token(EndOfLine, "\n", (25,23)->(25,23)),
        },
    ),
    NoOp(
        Token(EndOfLine, "\n", (26,0)->(26,0)),
    ),
]
